gitlab/resource_gitlab_project_environment.go

```diff
@@ -0,0 +1,163 @@
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+// https://docs.gitlab.com/ee/ci/environments/protected_environments.html
+func resourceGitlabProjectEnvironment() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGitlabProjectEnvironmentCreate,
+		Read:   resourceGitlabProjectEnvironmentRead,
+		Update: resourceGitlabProjectEnvironmentUpdate,
+		Delete: resourceGitlabProjectEnvironmentDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"external_url": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceGitlabProjectEnvironmentCreate(d *schema.ResourceData, meta interface{}) error {
+	name := d.Get("name").(string)
+	options := gitlab.CreateEnvironmentOptions{
+		Name: &name,
+	}
+	if externalURL, ok := d.GetOk("external_url"); ok {
+		options.ExternalURL = gitlab.String(externalURL.(string))
+	}
+
+	project := d.Get("project").(string)
+
+	log.Printf("[DEBUG] Project %s create gitlab environment %q", project, *options.Name)
+
+	client := meta.(*gitlab.Client)
+
+	environment, resp, err := client.Environments.CreateEnvironment(project, &options)
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			return fmt.Errorf("feature Environments is not available")
+		}
+		return err
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+
+	return resourceGitlabProjectEnvironmentRead(d, meta)
+}
+
+func resourceGitlabProjectEnvironmentRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] read gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return err
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+
+	log.Printf("[DEBUG] Project %s read gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	environment, resp, err := client.Environments.GetEnvironment(project, environmentIDInt)
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			log.Printf("[DEBUG] Project %s gitlab environment %q not found", project, environmentID)
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error getting gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+	d.Set("project", project)
+	d.Set("name", environment.Name)
+	d.Set("state", environment.State)
```

Comment:
I know it's currently done this way all over the provider, but it would be good to handle `error` for `Set`:

```suggestion
	err = d.Set("project", project)
	if err != nil {
		return err
	}

	err = d.Set("name", environment.Name)
	if err != nil {
		return err
	}

	err = d.Set("state", environment.State)
	if err != nil {
		return err
	}
```

Reply:
I'm gonna disagree here, as skipping error checking with primitive types is a regular practice in TF providers, since the SDK will catch these issues for us. Guidance for non-primitive types here: https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#error-checking-aggregate-types

Reply:
@armsnyder I suppose that goes for the rest of the suggested error checks, too?

Reply:
Yeah, but maybe wait until we come to some consensus in #651 😄